packages/cli/src/controllers/oauth/abstract-oauth.controller.ts

```diff
@@ -118,33 +125,67 @@ export abstract class AbstractOAuthController {
 		return await this.credentialsRepository.findOneBy({ id: credentialId });
 	}
 
-	createCsrfState(credentialsId: string): [string, string] {
+	createCsrfState(credentialsId: string, userId?: string): [string, string] {
 		const token = new Csrf();
 		const csrfSecret = token.secretSync();
 		const state: CsrfStateParam = {
 			token: token.create(csrfSecret),
 			cid: credentialsId,
+			createdAt: Date.now(),
+			userId,
 		};
 		return [csrfSecret, Buffer.from(JSON.stringify(state)).toString('base64')];
 	}
 
-	protected decodeCsrfState(encodedState: string): CsrfStateParam {
+	protected decodeCsrfState(encodedState: string, req: AuthenticatedRequest): CsrfStateParam {
 		const errorMessage = 'Invalid state format';
 		const decoded = jsonParse<CsrfStateParam>(Buffer.from(encodedState, 'base64').toString(), {
 			errorMessage,
 		});
+
 		if (typeof decoded.cid !== 'string' || typeof decoded.token !== 'string') {
 			throw new ApplicationError(errorMessage);
 		}
+
+		if (decoded.userId !== req.user?.id) {
+			throw new AuthError('Unauthorized');
+		}
+
 		return decoded;
 	}
 
 	protected verifyCsrfState(decrypted: ICredentialDataDecryptedObject, state: CsrfStateParam) {
 		const token = new Csrf();
-		return (
-			decrypted.csrfSecret === undefined ||
-			!token.verify(decrypted.csrfSecret as string, state.token)
+		if (decrypted.csrfSecret === undefined) return false;
+		if (!token.verify(decrypted.csrfSecret as string, state.token)) return false;
+		if (!state.createdAt || Date.now() - state.createdAt > 5 * Time.minutes.toMilliseconds)
```

Comment:
`state.createdAt` is `number` not `number | undefined`, is the type incorrect or the check unnecessary?

Reply:
The type is correct. I think I added the check assuming that there could be oauth requests in-flight when an instance upgrades, which might break the oauth flow. 
But that is probably not very likely, and the user can always try again.
I'll remove this.